### DISCUSSION TITLE
Search for paragraphs without specific patterns

### DISCUSSION THREAD

**Discussion by vbd** [BurntSushi/ripgrep#2077] [Q&A] [answered]
Input

    row 1
    row 2

    row 1

    row 1
    row 2
    row 3
    row 4
    row 5
    row 6

    row 1
    row 2
    row 3
    row 4
    row 5

How is it possible to search for a paragraph consisting only of 1 line, without a specific pattern of charachters?
Should return only

    row 1

And if it is possible how would I do so for a paragraph consisting excatly of 5 lines?
Should return

    row 1
    row 2
    row 3
    row 4
    row 5

I mix both cases here because I think the regex, if possible,  will look very differently.

Thanks in advance.

Update:
I tried something like: https://regex101.com/r/VXKAAN/1 but I'm really not good at regex.

**Comment by BurntSushi** [comment, marked as answer]
ripgrep is a line oriented tool. So it's generally not the right tool for this sort of task. By "line oriented tool," what I mean is that its basic mode of operation is to iterate over lines in a file and print only the ones that match. What you want to do requires something more powerful: the ability to reason about lines more holistically than that. ripgrep does not give you that arbitrary power.

With that said, ripgrep does have a `-U/--multiline` flag that will let you employ such reasoning in a limited way. To achieve what you want, for 1 and 5 lines, respectively:

```
$ rg -U '(?-m)(^|\n\n)(?P<paragraph>.+($|\n)){1}($|\n)' -o
input
4:row 1

$ rg -U '(?-m)(^|\n\n)(?P<paragraph>.+($|\n)){5}($|\n)' -o
input
13:row 1
14:row 2
15:row 3
16:row 4
17:row 5
```

EDIT: Obviously, this only works if you're using `\n` as line endings. But if you're on Windows, it's likely that your lines end with `\r\n` instead of just `\n`. So you'll need to replace `\n` with `\r\n` in the regex above. And possibly use the `--crlf` flag.

**Comment by vbd** [comment]
At first thank you very much for your help.

I tested it my windows pc but It didn't work. I tried it in CMD and Powershell. Did replace the ' with " for first debugging. 
In powershell:
`rg -U '(?-m)(^|\n\n)(?P<paragraph>.+($|\n)){1}($|\n)' -o input`
returns no error, no result

In CMD:
`rg -U '(?-m)(^|\n\n)(?P<paragraph>.+($|\n)){1}($|\n)' -o input`
returns "file can't be found" (in german: Das System kann die angegebene Datei nicht finden.)

`rg -U "(?-m)(^|\n\n)(?P<paragraph>.+($|\n)){1}($|\n)" -o input`
returns no error, no result

Any idea what I am missing?

**Reply by BurntSushi** [reply]
Well, you need to make sure `input` is a real file... Otherwise, I don't know, I don't use Windows. Sorry. It works for me.
![Uploading rg-2077-works.png…]()

**Reply by vbd** [reply]
input is a real file. 
Just tested it on cygwin bash, also did not work.

File is named input, to be able to upload I named it input.txt.

[input.txt](https://github.com/BurntSushi/ripgrep/files/7555614/input.txt)

Thanks for your help anyway.

**Reply by BurntSushi** [reply]
Sigh. Windows uses `\r\n` for line endings. Not `\n`. So you need `'(?-m)(^|\r\n\r\n)(?P<paragraph>.+($|\r\n)){5}($|\r\n)'`. And maybe the `--crlf` flag to handle some edge cases.

**Reply by vbd** [reply]
I am so stupid, I have overlooked it once again ... I created a file for testing purposes on my windows box. The source files are from a linux machine.
Thank you so so so much. It works like a charm.
